File: battle-of-heroes/game/tools/playtest.py
from pathlib import Path
from playwright.sync_api import sync_playwright

out = Path(__file__).resolve().parent.parent / "qa"
out.mkdir(exist_ok=True)
errors: list[str] = []

with sync_playwright() as p:
    browser = p.chromium.launch(headless=True)
    page = browser.new_page(viewport={"width": 1280, "height": 720})
    page.on("pageerror", lambda e: errors.append(str(e)))
    page.on("console", lambda m: errors.append(m.text) if m.type == "error" else None)
    page.goto("http://127.0.0.1:5173/", wait_until="networkidle", timeout=120000)
    page.wait_for_selector("canvas", timeout=60000)
    page.wait_for_timeout(2800)
    page.screenshot(path=str(out / "01-menu.png"))

    canvas = page.locator("canvas")
    box = canvas.bounding_box()
    assert box

    def click(x: float, y: float) -> None:
        page.mouse.click(box["x"] + x, box["y"] + y)

    click(470, 338)
    page.wait_for_timeout(900)
    page.screenshot(path=str(out / "02-map.png"))

    click(1280 * 0.17, 720 * 0.78)
    page.wait_for_timeout(2200)
    page.screenshot(path=str(out / "03-battle.png"))

    for i in range(10):
        click(640 + (i % 5 - 2) * 118, 662)
        page.wait_for_timeout(280)
    page.wait_for_timeout(4500)
    page.screenshot(path=str(out / "04-combat.png"))

    click(30, 22)
    page.wait_for_timeout(700)
    click(90, 40)
    page.wait_for_timeout(500)
    click(810, 338)
    page.wait_for_timeout(900)
    page.screenshot(path=str(out / "05-heroes.png"))

    click(90, 36)
    page.wait_for_timeout(500)
    click(470, 488)
    page.wait_for_timeout(900)
    page.screenshot(path=str(out / "06-shop.png"))

    print("errors:", errors)
    browser.close()

if errors:
    raise SystemExit(2)

File: battle-of-heroes/game/tools/playtest.mjs
import { chromium } from "playwright";
import { mkdirSync } from "fs";
import { dirname, join } from "path";
import { fileURLToPath } from "url";

const dir = join(dirname(fileURLToPath(import.meta.url)), "..", "qa");
mkdirSync(dir, { recursive: true });

const browser = await chromium.launch({ headless: true });
const page = await browser.newPage({ viewport: { width: 1280, height: 720 } });
const errors = [];
page.on("pageerror", (e) => errors.push(String(e)));
page.on("console", (msg) => {
  if (msg.type() === "error") errors.push(msg.text());
});

await page.goto("http://127.0.0.1:5173/", { waitUntil: "networkidle", timeout: 120000 });
await page.waitForSelector("canvas", { timeout: 60000 });
await page.waitForTimeout(2500);
await page.screenshot({ path: join(dir, "01-menu.png") });

const canvas = page.locator("canvas");
const box = await canvas.boundingBox();
if (!box) throw new Error("no canvas");
const click = async (x, y) => {
  await page.mouse.click(box.x + x, box.y + y);
};

await click(470, 338);
await page.waitForTimeout(800);
await page.screenshot({ path: join(dir, "02-map.png") });

await click(1280 * 0.13, 720 * 0.7);
await page.waitForTimeout(2000);
await page.screenshot({ path: join(dir, "03-battle.png") });

for (let i = 0; i < 8; i++) {
  await click(640 + (i % 5 - 2) * 118, 662);
  await page.waitForTimeout(350);
}
await page.waitForTimeout(4000);
await page.screenshot({ path: join(dir, "04-combat.png") });

await click(24, 18);
await page.waitForTimeout(600);
await click(90, 40);
await page.waitForTimeout(500);
await click(810, 338);
await page.waitForTimeout(800);
await page.screenshot({ path: join(dir, "05-heroes.png") });

await click(90, 36);
await page.waitForTimeout(400);
await click(470, 488);
await page.waitForTimeout(800);
await page.screenshot({ path: join(dir, "06-shop.png") });

const pixels = await page.evaluate(() => {
  const c = document.querySelector("canvas");
  if (!c) return { empty: true };
  const ctx = c.getContext("2d") || c.getContext("webgl") || c.getContext("webgl2");
  return {
    w: c.width,
    h: c.height,
    webgl: !!(c.getContext("webgl") || c.getContext("webgl2")),
  };
});

console.log(JSON.stringify({ errors, pixels }, null, 2));
await browser.close();
if (errors.length) process.exitCode = 2;
